### PR TITLE
perf(common): faster logs by caching intl.datetimeformat

### PR DESCRIPTION
**Pull request by H4ad**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?

This is the profiler information when we call `#log` method 10.000 times:

```
[Bottom up (heavy) profile]:
  Note: percentage shows a share of a particular caller in the total
  amount of its parent calls.
  Callers occupying less than 1.0% are not shown.

   ticks parent  name
    958   56.8%  /home/h4ad/.asdf/installs/nodejs/12.22.12/bin/node
    763   79.6%    v8::internal::Builtin_DatePrototypeToLocaleString(int, unsigned long*, v8::internal::Isolate*)
    560   73.4%      LazyCompile: *getTimestamp /home/h4ad/Projects/opensource/performance-test-nestjs/test-performance/node_modules/@nestjs/common/services/console-logger.service.js:104:17
    510   91.1%        LazyCompile: *log /home/h4ad/Projects/opensource/performance-test-nestjs/test-performance/node_modules/@nestjs/common/services/console-logger.service.js:28:8
    510  100.0%          LazyCompile: *log /home/h4ad/Projects/opensource/performance-test-nestjs/test-performance/node_modules/@nestjs/common/services/logger.service.js:36:8
    510  100.0%            LazyCompile: ~descriptor.value /home/h4ad/Projects/opensource/performance-test-nestjs/test-performance/node_modules/@nestjs/common/services/logger.service.js:155:33
     50    8.9%        LazyCompile: *printMessages /home/h4ad/Projects/opensource/performance-test-nestjs/test-performance/node_modules/@nestjs/common/services/console-logger.service.js:115:18
     50  100.0%          LazyCompile: ~log /home/h4ad/Projects/opensource/performance-test-nestjs/test-performance/node_modules/@nestjs/common/services/console-logger.service.js:28:8
     50  100.0%            LazyCompile: *log /home/h4ad/Projects/opensource/performance-test-nestjs/test-performance/node_modules/@nestjs/common/services/logger.service.js:36:8
    203   26.6%      LazyCompile: ~getTimestamp /home/h4ad/Projects/opensource/performance-test-nestjs/test-performance/node_modules/@nestjs/common/services/console-logger.service.js:104:17
    179   88.2%        LazyCompile: ~formatMessage /home/h4ad/Projects/opensource/performance-test-nestjs/test-performance/node_modules/@nestjs/common/services/console-logger.service.js:131:18
    179  100.0%          LazyCompile: ~<anonymous> /home/h4ad/Projects/opensource/performance-test-nestjs/test-performance/node_modules/@nestjs/common/services/console-logger.service.js:116:26
    179  100.0%            /home/h4ad/.asdf/installs/nodejs/12.22.12/bin/node
     24   11.8%        LazyCompile: *printMessages /home/h4ad/Projects/opensource/performance-test-nestjs/test-performance/node_modules/@nestjs/common/services/console-logger.service.js:115:18
     24  100.0%          LazyCompile: ~log /home/h4ad/Projects/opensource/performance-test-nestjs/test-performance/node_modules/@nestjs/common/services/console-logger.service.js:28:8
     24  100.0%            LazyCompile: ~log /home/h4ad/Projects/opensource/performance-test-nestjs/test-performance/node_modules/@nestjs/common/services/logger.service.js:36:8
```

> [Source](https://github.com/nestjs/nest/files/10366481/isolate-0x3b00030-34045-v8_normal_logger.log)

As you can see, `getTimestamp` consumes most of the execution time, which doesn't make sense, I did some CPU profiling and I found this:

image

> [normal_logger_nodejs12.cpuprofile.zip](https://github.com/nestjs/nest/files/10366531/normal_logger_nodejs12.cpuprofile.zip)

The `#getTimestamp` method took 1ms, in another test that I did, this function

When we run again on nodejs16 or nodejs19:

image

The `#getTimestamp` method took 13ms (I think I found a regression on NodeJS).

> [normal_logger_nodejs16.cpuprofile.zip](https://github.com/nestjs/nest/files/10366528/normal_logger_nodejs16.cpuprofile.zip)
> [normal_logger_nodejs19.cpuprofile.zip](https://github.com/nestjs/nest/files/10366525/normal_logger_nodejs19.cpuprofile.zip)

Also, decorators like `WrapBuffer` tends to not be optimized by V8, so this could be another point of optimization.

Issue Number: N/A

## What is the new behavior?

So, to solve the issue with `#getTimestamp`, I started using `Intl.DateTimeFormat` API,
and for the decorator `WrapBuffer`, I just refactor to use a function.

The final profiler information became:

```
[Bottom up (heavy) profile]:
  Note: percentage shows a share of a particular caller in the total
  amount of its parent calls.
  Callers occupying less than 1.0% are not shown.

   ticks parent  name
    207   45.3%  /home/h4ad/.asdf/installs/nodejs/12.22.12/bin/node
     61   29.5%    /home/h4ad/.asdf/installs/nodejs/12.22.12/bin/node
      6    9.8%      /home/h4ad/.asdf/installs/nodejs/12.22.12/bin/node
      2   33.3%        LazyCompile: ~readPackage internal/modules/cjs/loader.js:231:21
      2  100.0%          LazyCompile: ~readPackageScope internal/modules/cjs/loader.js:262:26
      1   50.0%            LazyCompile: ~trySelf internal/modules/cjs/loader.js:383:17
      1   50.0%            LazyCompile: ~Module._extensions..js internal/modules/cjs/loader.js:1008:37
      2   33.3%        LazyCompile: ~Module._findPath internal/modules/cjs/loader.js:436:28
      2  100.0%          LazyCompile: ~Module._resolveFilename internal/modules/cjs/loader.js:731:35
      2  100.0%            LazyCompile: ~Module._load internal/modules/cjs/loader.js:648:24
      1   16.7%        LazyCompile: ~stat internal/modules/cjs/loader.js:123:14
      1  100.0%          LazyCompile: ~tryFile internal/modules/cjs/loader.js:326:17
      1  100.0%            LazyCompile: ~tryExtensions internal/modules/cjs/loader.js:342:23
      1   16.7%        LazyCompile: ~realpathSync fs.js:1568:22
      1  100.0%          LazyCompile: ~toRealPath internal/modules/cjs/loader.js:335:20
      1  100.0%            LazyCompile: ~tryFile internal/modules/cjs/loader.js:326:17
      5    8.2%      LazyCompile: ~toRealPath internal/modules/cjs/loader.js:335:20
      4   80.0%        LazyCompile: ~tryFile internal/modules/cjs/loader.js:326:17
      4  100.0%          LazyCompile: ~tryExtensions internal/modules/cjs/loader.js:342:23
      3   75.0%            LazyCompile: ~Module._findPath internal/modules/cjs/loader.js:436:28
      1   25.0%            LazyCompile: ~tryPackage internal/modules/cjs/loader.js:279:20
      1   20.0%        LazyCompile: ~Module._findPath internal/modules/cjs/loader.js:436:28
      1  100.0%          LazyCompile: ~resolveMainPath internal/modules/run_main.js:8:25
      1  100.0%            LazyCompile: ~executeUserEntryPoint internal/modules/run_main.js:53:31
      5    8.2%      LazyCompile: *log /home/h4ad/Projects/opensource/performance-test-nestjs/test-performance/dist/perf/better-console-logger.service.js:47:8
      4   80.0%        LazyCompile: *log /home/h4ad/Projects/opensource/performance-test-nestjs/test-performance/dist/perf/better-logger.service.js:52:8
      4  100.0%          Eval: ~<anonymous> /home/h4ad/Projects/opensource/performance-test-nestjs/test-performance/dist/perf/perf-better-logger.js:1:1
      4  100.0%            LazyCompile: ~Module._compile internal/modules/cjs/loader.js:953:37
      1   20.0%        LazyCompile: *<anonymous> /home/h4ad/Projects/opensource/performance-test-nestjs/test-performance/dist/perf/perf-better-logger.js:1:1
      1  100.0%          LazyCompile: ~Module._compile internal/modules/cjs/loader.js:953:37
      1  100.0%            LazyCompile: ~Module._extensions..js internal/modules/cjs/loader.js:1008:37
```

> [Source](https://github.com/nestjs/nest/files/10366508/perf-better-logger.txt)

And the CPU profiling now looks like:

image

And the time to perform the log looks consistent between NodeJS environments.

> [perf-better-nodejs12.cpuprofile.zip](https://github.com/nestjs/nest/files/10366541/perf-better-nodejs12.cpuprofile.zip)
> [perf-better-nodejs16.cpuprofile.zip](https://github.com/nestjs/nest/files/10366542/perf-better-nodejs16.cpuprofile.zip)
> [perf-better-nodejs19.cpuprofile.zip](https://github.com/nestjs/nest/files/10366543/perf-better-nodejs19.cpuprofile.zip)

Finally, about the performance improvements, this is the result running on NodeJS 16:

```
Logger x 7,127 ops/sec ±1.55% (87 runs sampled)
BetterLogger x 60,055 ops/sec ±1.90% (65 runs sampled)
Console x 133,689 ops/sec ±1.14% (90 runs sampled)
```

> Console is just `console.log`.

An improvement of 8x in the performance when logging.

I didn't make any breaking changes in the logs, but if we change from `Intl.DateTimeFormat` API to just `new Date().toISOString()`, the benchmark will look like:

```
Logger x 7,165 ops/sec ±1.80% (80 runs sampled)
BetterLogger x 79,513 ops/sec ±2.74% (62 runs sampled)
Console x 140,275 ops/sec ±1.61% (88 runs sampled)
```

So let me know if worth to introduce some breaking changes just to have better performance when logging.

To reproduce all those tests, see this gist: https://gist.github.com/H4ad/da838dfc33b2060504ca980644d804b5

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

## Other information

### CODE REVIEW

**Comment by micalevisk** [comment]
can we split those two optizations into 2 separated commits

**Comment by H4ad** [comment]
@micalevisk  Done!

**Review by micalevisk — APPROVED** [review]

**Review by Tony133 — APPROVED** [review]

File: packages/common/services/logger.service.ts [hunk 1]
```diff
@@ -97,73 +76,89 @@ export class Logger implements LoggerService {
...
-      ? optionalParams.concat(this.context)
-      : optionalParams;
+    Logger.wrapBuffer(() => {
+      optionalParams = this.context
+        ? optionalParams.concat(this.context)
+        : optionalParams;
 
-    this.localInstance?.error(message, ...optionalParams);
+      this.localInstance?.error(message, ...optionalParams);
+    });
```
Comment (kamilmysliwiec) [inline packages/common/services/logger.service.ts:101]: Could you please revert this change and show what diff would that cause in performance? Decorators-driven approach simply looks much cleaner

Comment (H4ad) [reply packages/common/services/logger.service.ts:101]: I run the tests three times:

```
'Logger x 6,629 ops/sec ±2.15% (82 runs sampled)',
'BetterLogger x 59,884 ops/sec ±2.88% (72 runs sampled)',
'BetterLoggerWithDecoratorLogger x 61,333 ops/sec ±2.57% (72 runs sampled)',
'Console x 134,305 ops/sec ±1.28% (86 runs sampled)',
'process.stdout x 299,742 ops/sec ±1.75% (75 runs sampled)'

'Logger x 6,702 ops/sec ±1.86% (86 runs sampled)',
'BetterLogger x 58,865 ops/sec ±2.12% (71 runs sampled)',
'BetterLoggerWithDecoratorLogger x 58,864 ops/sec ±1.85% (69 runs sampled)',
'Console x 140,193 ops/sec ±1.33% (88 runs sampled)',
'process.stdout x 284,481 ops/sec ±1.26% (68 runs sampled)'

'Logger x 6,529 ops/sec ±1.58% (85 runs sampled)',
'BetterLogger x 58,534 ops/sec ±2.24% (67 runs sampled)',
'BetterLoggerWithDecoratorLogger x 60,301 ops/sec ±2.53% (69 runs sampled)',
'Console x 137,276 ops/sec ±0.94% (87 runs sampled)',
'process.stdout x 306,507 ops/sec ±1.95% (80 runs sampled)'
```

Is almost the same or a little bit faster with decorators (I'm very surprised with this result).

Looking in the profiler information:

With decorator:

```
Statistical profiling result from perf-better-logger-with-decorator.log, (1911 ticks, 0 unaccounted, 0 excluded).

 [Shared libraries]:
   ticks  total  nonlib   name
    722   37.8%          /home/h4ad/.asdf/installs/nodejs/12.22.12/bin/node
     26    1.4%          /usr/lib/x86_64-linux-gnu/libc-2.31.so
      8    0.4%          [vdso]

 [JavaScript]:
   ticks  total  nonlib   name
      8    0.4%    0.7%  LazyCompile: *resolve path.js:973:10
      8    0.4%    0.7%  LazyCompile: *formatMessage /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/dist/perf/better-console-logger.service.js:131:18
      8    0.4%    0.7%  LazyCompile: *<anonymous> /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/dist/perf/perf-better-logger.js:1:1
      6    0.3%    0.5%  LazyCompile: *Writable.write _stream_writable.js:291:36
      5    0.3%    0.4%  LazyCompile: *normalizeString path.js:52:25
      3    0.2%    0.3%  LazyCompile: *createWriteWrap internal/stream_base_commons.js:106:25
      2    0.1%    0.2%  LazyCompile: *updateAndGetTimestampDiff /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/dist/perf/better-console-logger.service.js:152:30
      2    0.1%    0.2%  LazyCompile: *get internal/bootstrap/pre_execution.js:297:8
      1    0.1%    0.1%  LazyCompile: *log /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/dist/perf/better-console-logger.service.js:47:8
      1    0.1%    0.1%  LazyCompile: *handleWriteReq internal/stream_base_commons.js:47:24
      1    0.1%    0.1%  LazyCompile: *getStdout internal/bootstrap/switches/is_main_thread.js:115:19
      1    0.1%    0.1%  LazyCompile: *formatContext /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/dist/perf/better-console-logger.service.js:128:18
```

Without decorator:

```
Statistical profiling result from perf-better-logger.log, (1956 ticks, 2 unaccounted, 0 excluded).

 [Shared libraries]:
   ticks  total  nonlib   name
    720   36.8%          /home/h4ad/.asdf/installs/nodejs/12.22.12/bin/node
     24    1.2%          /usr/lib/x86_64-linux-gnu/libc-2.31.so
      5    0.3%          [vdso]
      1    0.1%          /usr/lib/x86_64-linux-gnu/libm-2.31.so

 [JavaScript]:
   ticks  total  nonlib   name
     13    0.7%    1.1%  LazyCompile: *log /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/dist/perf/better-console-logger.service.js:47:8
      7    0.4%    0.6%  LazyCompile: *writeOrBuffer _stream_writable.js:355:23
      5    0.3%    0.4%  LazyCompile: *getTimestamp /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/dist/perf/better-console-logger.service.js:112:17
      4    0.2%    0.3%  LazyCompile: *resolve path.js:973:10
      3    0.2%    0.2%  LazyCompile: *validChunk _stream_writable.js:275:20
      3    0.2%    0.2%  LazyCompile: *stringifyMessage /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/dist/perf/better-console-logger.service.js:137:21
      3    0.2%    0.2%  LazyCompile: *formatContext /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/dist/perf/better-console-logger.service.js:128:18
      2    0.1%    0.2%  LazyCompile: *normalizeString path.js:52:25
      2    0.1%    0.2%  LazyCompile: *get internal/bootstrap/pre_execution.js:297:8
      2    0.1%    0.2%  LazyCompile: *<anonymous> /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/dist/perf/perf-better-logger.js:1:1
      1    0.1%    0.1%  LazyCompile: *log /home/h4ad/Projects/opensource/performance-analysis/test-performance-nestjs/dist/perf/better-logger.service.js:53:8
      1    0.1%    0.1%  LazyCompile: *getStdout internal/bootstrap/switches/is_main_thread.js:115:19
      1    0.1%    0.1%  LazyCompile: *Module._nodeModulePaths internal/modules/cjs/loader.js:567:37
```

With decorator has 45 ticks less than the version without decorator.

I will remove the commit and leave just the `Intl` commit.

**Comment by H4ad** [comment]
Based on https://github.com/jmcdo29/ogma/pull/1493, If we use the library [sonic-boom](https://github.com/pinojs/sonic-boom), I think we can achieve 200k~300k/ops.

In ogma, we went from 100k to 400k op/s.

Sonic boom is part of PinoJS and it buffers the logs, then sends to directly to file descriptor, using this approach is faster than writing directly to `process.stdout`.

Do you think it's worth optimizing?

**Comment by kamilmysliwiec** [comment]
Sounds good @H4ad. Would you like to create a separate PR for this?
